Added dynamic reconfigure, auto white balance

lwan-zz opened the pull request:
Exposed current camera settings to dynamic reconfigure: 
- Gain (and auto gain toggle)
- Brightness (and auto brightness toggle -- neither seem to do anything to the image)
- Exposure (and auto exposure)
- Gamma
- Auto white balance enable

Other settings will need additional hooks into the Pylon API (custom white balance setttings, trigger settings, etc.)

Also set default MTU from 3000 to 1500 in case it's not set in the .launch file.

lwan committed:
added auto white balance wrapper; things broke

lwan committed:
Added auto white balance

Camera driver now enables "single" auto white balance on bringup.

lwan committed:
Setup dynamic_reconfigure
Still needs to be integrated with rest of camera paramters

lwan committed:
Split grab image into init, set params
Allows for use with dynamic reconfigure

lwan committed:
Exposed current camera setting to dynamic reconfig

Exposure, gain, gamma, brightness (and auto toggles thereof) are exposed
to dynamic reconfigure. Will need to add other Pylon API hooks for other
settings.

jonbinney commented on package.xml:
```diff
@@ -47,8 +47,9 @@
-  <!--run_depend>std_srvs</run_depend-->
   <run_depend>nodelet</run_depend>
+  <run_depend>dynamic_reconfigure</run_depend>
+  <run_depend>roslint</run_depend>
```
Why did you make roslint a run_depend?

lwan-zz replied:
Bad habit to make build_depends also run_depends. Will take out.

lwan-zz replied:
@jonbinney doesn't look like it'll build without roslint as a run_depend:

`CMake Error at /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:224 (message):
  catkin_package() DEPENDS on the catkin package 'roslint' which must
  therefore be listed as a run dependency in the package.xml
Call Stack (most recent call first):
  /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:102 (_catkin_package)
  pylon_camera/CMakeLists.txt:40 (catkin_package)
`

jonbinney commented on src/pylon_camera/main.cpp:
```diff
@@ -35,19 +35,34 @@
     ros::NodeHandle nh_image;
 
+	dynamic_reconfigure::Server<pylon_camera::CameraSettingsConfig> server;
+    dynamic_reconfigure::Server<pylon_camera::CameraSettingsConfig>::CallbackType f;
```
This shouldn't be in `main()` now, right? You moved it to the PylonCameraNode class.

lwan-zz replied:
Yep. Taking it out.

jonbinney commented on src/pylon_camera/main.cpp:
```diff
@@ -35,19 +35,34 @@
     ROS_INFO_STREAM("Start image grabbing if node connects to topic with "
         << "a frame_rate of: " << pylon_camera_node.frameRate() << " Hz");
-
+   
```
Remove white space. Actually `main.cpp` should probably not have any changes at all in this PR.

lwan-zz replied:
Will do. Reverting changes to `main.cpp`

jonbinney commented on src/pylon_camera/pylon_camera_node.cpp:
```diff
@@ -34,6 +34,7 @@
 #include <vector>
 #include "boost/multi_array.hpp"
 
+
```
Try not to make whitespace changes in a feature PR unless there is a strong reason to. They make it harder to find the real changes in the diff.

lwan-zz replied:
Will remove.

jonbinney commented on src/pylon_camera/pylon_camera_node.cpp:
```diff
@@ -178,14 +222,13 @@ bool PylonCameraNode::startGrabbing()
-                                                        _1,
-                                                        _2));
+        set_user_output_srvs_.at(i) = nh_private_->advertiseService< camera_control_msgs::SetBool::Request, 
+                                                                     camera_control_msgs::SetBool::Response >( 
```
Is this just a whitespace change?

jonbinney replied:
Actually, screw backwards compatibility in this case. Can you just get rid of the individual parameter setting services? They are obsolete once we can change things with dynamic_reconfigure.

lwan-zz replied:
Yep. I'll remove the services per our conversation today.

jonbinney commented on src/pylon_camera/pylon_camera_parameter.cpp:
```diff
@@ -204,6 +207,16 @@ void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
+
+
+    
+    if ( nh.hasParam("whitebalance_auto") )
```
Dynamic reconfigure should call your callback with the values set as ROS params at startup - is there a reason you also check the parameter manually here?

lwan-zz replied:
@jonbinney:  
Wouldn't I want to see if the parameter is already set? Dynamic reconfigure should be setting it at startup, but what if the launchfile sets a different value?

jonbinney replied:
Dynamic reconfigure actually does this. It checks for a ROS parameter with the name of the parameter at startup. So if you set it in the launchfile, dynamic reconfigure will pass that value to your callback.

jonbinney commented on src/pylon_camera/pylon_camera_parameter.cpp:
```diff
@@ -71,6 +72,8 @@ PylonCameraParameter::~PylonCameraParameter()
 
 void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
 {
+    ROS_INFO_STREAM("Reading parameters from paramter server");
```
Spelling: `paramter`

lwan-zz replied:
Okey doke. Will change.

jonbinney commented on src/pylon_camera/pylon_camera_parameter.cpp:
```diff
@@ -45,7 +45,7 @@ PylonCameraParameter::PylonCameraParameter() :
         binning_y_given_(false),
         downsampling_factor_exp_search_(1),
```
Can we get rid of this `PylonCameraParameter` class completely and replace it with the one that dynamic-reconfigure generates automatically? That will require adding the remaining parameters to dynamic_reconfigure (e.g. `mtu_size`), but that will be a good idea anyway.

lwan-zz replied:
Yep, this should be removed as part of moving everything to dynamic reconfigure.

jonbinney commented on src/pylon_camera/pylon_camera_parameter.cpp:
```diff
@@ -204,6 +207,16 @@ void PylonCameraParameter::readFromRosParameterServer(const ros::NodeHandle& nh)
+    if ( nh.hasParam("whitebalance_auto") )
+    {
+        nh.getParam("whitebalance_auto", whitebalance_auto_);
+        std::cout << "continuous auto white balance set" << std::endl;
```
Don't use `cout` here. If you need to output, use `LOG_INFO`.

lwan-zz replied:
Will do.

lwan-zz replied:
Any reason not to use `ROS_INFO_STREAM` ?

jonbinney replied:
ROS_INFO_STREAM is fine too - as long as the output goes through the ROS logging system.

lwan committed:
Updated changes per pull request

Also increased default framerate from 5 -> 30, set default global
shutter.

`PylonCameraParameter` was not removed, since it proved to be pretty
useful (for now).

jonbinney commented:
Looks good!
